ykcompile/src/lib.rs

```diff
@@ -36,44 +36,52 @@ pub struct TraceCompiler {
     /// Contains the list of currently available registers.
     available_regs: Vec<u8>,
     /// Maps locals to their assigned registers.
-    assigned_regs: HashMap<u32, u8>,
+    assigned_regs: HashMap<Local, u8>,
 }
 
 impl TraceCompiler {
-    fn local_to_reg(&mut self, l: u32) -> u8 {
+    fn local_to_reg(&mut self, l: Local) -> u8 {
         // This is a really dumb register allocator, which runs out of available registers after 7
         // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
         // up registers again, and allocate additional locals on the stack. Though, ultimately we
         // probably want to implement a proper register allocator, e.g. linear scan.
-        if l == 0 {
+
+        if l == Local(0) {
+            // In SIR, `Local` zero is the (implicit) return value, so it makes sense to allocate
+            // it in return register of the underlying X86_64 calling convention.
             0
         } else {
-            let reg = self
-                .available_regs
-                .pop()
-                .expect("Can't allocate more than 7 locals yet!");
-            *self.assigned_regs.entry(l).or_insert(reg)
+            if self.assigned_regs.contains_key(&l) {
+                self.assigned_regs[&l]
+            } else {
+                let reg = self.available_regs.pop().expect(
+                    "Can't allocate more than 7 locals
+                    yet!",
+                );
+                self.assigned_regs.insert(l, reg);
+                reg
+            }
```

Comment:
```suggestion
            let available_regs = &mut self.available_regs;
            *self.assigned_regs.entry(l).or_insert_with(|| {
                available_regs.pop().expect("Can't allocate more than 7 locals yet!");
            })
```

Reply:
I tried something similar to this and it whined about mutable borrows inside the `or_insert_with` closure. Let's give it a shot.

Reply:
Worked with a minor tweak :)

I think the trick is to do `let available_regs = &mut self.available_regs`.

I had repeated `self.available_regs` inside the closure, which upset the borrow checker.

Thanks!

Reply:
> I think the trick is to do let available_regs = &mut self.available_regs.

See my code :) A closure currently captures a whole variable (`self`), not just a part of it (`self.available_regs`). There is an accepted RFC to only capture the used part, but the implementation is currently on hold.

Reply:
https://github.com/rust-lang/rfcs/pull/2229